# CSE 151B: Linear Algebra and Calculus Review

In [ ]:
# import PyTorch
import torch

## 1. Linear Algebra

### Scalars, Vectors, Matrices, and Tensors

A Scalar $\mathbf{x} \in \mathbb{R}$ is a real number.

A vector $\mathbf{y} \in \mathbb{R}^n = \begin{pmatrix}\mathbf{y}_1\\\mathbf{y}_2\\\vdots\\\mathbf{y}_n\end{pmatrix}$ is a fixed-length array of scalars.

A matrix $\mathbf{z} \in \mathbb{R}^{m \times n} = \begin{pmatrix}\mathbf{z}_{11} && \mathbf{z}_{12} && \dots && \mathbf{z}_{1n}\\\mathbf{z}_{21} && \mathbf{z}_{22} && \dots && \mathbf{z}_{2n}\\\vdots && \vdots && \ddots && \vdots\\\mathbf{z}_{m1} && \mathbf{z}_{m2} && \dots && \mathbf{z}_{mn}\end{pmatrix}$ is a second-order tensor (or an array or arrays if you will) consisting of two axes---rows and columns.

A tensor extends this idea into higher dimensions; it is an $n^{th}$-order array.

For example,  
- A single grayscale image (having a single channel) is represented as a second-order tensor: $\mathbf{H} \times \mathbf{W}$
- A single colored image (having 3 channels RGB) is represented as a third-order tensor: $3 \times \mathbf{H} \times \mathbf{W}$
- A batch (collection) of grayscale images is a third-order tensor: $\mathbf{N} \times \mathbf{H} \times \mathbf{W}$
- A batch (collection) of colored images is a fourth-order tensor: $\mathbf{N} \times 3 \times \mathbf{H} \times \mathbf{W}$


In [ ]:
x = torch.tensor(3.0)
print("A scalar x:", x)

y = torch.arange(5)
print("\nA vector y:", y)
print("Length of vector y:", len(y))
print("2nd element of vector y:", y[1])  # zero-based indexing

z = torch.arange(6).reshape(3, 2)
print("\nA matrix z:\n", z)
print("Number of rows:", z.shape[0])
print("Number of columns:", z.shape[1])
print("Element at position (2,1):", z[2][1])
print("Transpose of matrix z:\n", z.T)

w = torch.arange(12).reshape(2,2,3)
print("\nA 3rd-order tensor w:\n", w)
print("Shape of w:", w.shape)

A scalar x: tensor(3.)

A vector y: tensor([0, 1, 2, 3, 4])
Length of vector y: 5
2nd element of vector y: tensor(1)

A matrix z:
 tensor([[0, 1],
        [2, 3],
        [4, 5]])
Number of rows: 3
Number of columns: 2
Element at position (2,1): tensor(5)
Transpose of matrix z:
 tensor([[0, 2, 4],
        [1, 3, 5]])

A 3rd-order tensor w:
 tensor([[[ 0,  1,  2],
         [ 3,  4,  5]],

        [[ 6,  7,  8],
         [ 9, 10, 11]]])
Shape of w: torch.Size([2, 2, 3])


### Elementwise Operations

In [ ]:
a = torch.arange(6).reshape(3,2)
b = a.clone()
print("a:\n", a)
print("b:\n", b)
print("Elementwise Addition:\n", a + b)
print("Elementwise Multiplication:\n", a * b)
print("Elementwise Function Application:\n", torch.square(a))

a:
 tensor([[0, 1],
        [2, 3],
        [4, 5]])
b:
 tensor([[0, 1],
        [2, 3],
        [4, 5]])
Elementwise Addition:
 tensor([[ 0,  2],
        [ 4,  6],
        [ 8, 10]])
Elementwise Multiplication:
 tensor([[ 0,  1],
        [ 4,  9],
        [16, 25]])
Elementwise Function Application:
 tensor([[ 0,  1],
        [ 4,  9],
        [16, 25]])


### Reduction

For a matrix $\mathbf{z} \in \mathbb{R}^{m \times n} = \begin{pmatrix}\mathbf{z}_{11} && \mathbf{z}_{12} && \dots && \mathbf{z}_{1n}\\\mathbf{z}_{21} && \mathbf{z}_{22} && \dots && \mathbf{z}_{2n}\\\vdots && \vdots && \ddots && \vdots\\\mathbf{z}_{m1} && \mathbf{z}_{m2} && \dots && \mathbf{z}_{mn}\end{pmatrix}$,

In [ ]:
z = torch.arange(10).reshape(5,2)
print("z:\n", z)

z:
 tensor([[0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]])


[A] $\sum_{i=0}^m \sum_{j=0}^n \mathbf{z}[i][j]$

In [ ]:
print(torch.sum(z))

tensor(45)


[B] $\sum_{i=0}^m \mathbf{z}[i][j]$  $\forall j \in \{0, \dots, n\}$

In [ ]:
print(torch.sum(z, axis=0))

tensor([20, 25])


[C] $\sum_{j=0}^n \mathbf{z}[i][j]$  $\forall i \in \{0, \dots, m\}$

In [ ]:
print(torch.sum(z, axis=1))

tensor([ 1,  5,  9, 13, 17])


### Non-Reduction

Notice how, in the above examples, the result of `torch.sum` does not maintain the original shape of the tensor. Everything that is output is converted into a single-axis vector.

If we want to preserve the axes, pass the `keepdim=True` argument.

In [ ]:
print(torch.sum(z, axis=0, keepdim=True))  # should maintain the 2 columns
print(torch.sum(z, axis=1, keepdim=True))  # should maintain the 5 rows

tensor([[20, 25]])
tensor([[ 1],
        [ 5],
        [ 9],
        [13],
        [17]])


The relevance of `keepdim=True` comes into play when you want to broadcast a tensor.

Consider performing the mean (or average) operation along every row of a matrix:

In [ ]:
z = torch.arange(10).reshape(5,2)
print("z:\n", z)

sum = torch.sum(z, axis=1)
print("sum along every row:", sum)

print("mean along every row:\n", z / sum)

z:
 tensor([[0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]])
sum along every row: tensor([ 1,  5,  9, 13, 17])


RuntimeError: ignored

Now, use `keepdim=True` to preserve axes:

In [ ]:
z = torch.arange(10).reshape(5,2)
print("z:\n", z)

sum = torch.sum(z, axis=1, keepdim=True)
print("sum along every row:\n", sum)

print("mean along every row:\n", z / sum)

z:
 tensor([[0, 1],
        [2, 3],
        [4, 5],
        [6, 7],
        [8, 9]])
sum along every row:
 tensor([[ 1],
        [ 5],
        [ 9],
        [13],
        [17]])
mean along every row:
 tensor([[0.0000, 1.0000],
        [0.4000, 0.6000],
        [0.4444, 0.5556],
        [0.4615, 0.5385],
        [0.4706, 0.5294]])


Voila! What is happening behind the scenes is:

$\mathbf{z} = \begin{pmatrix}0 && 1\\2 && 3\\4&&5\\6 && 7\\8 && 9\end{pmatrix}$

$\text{sum} = \begin{pmatrix}1\\5\\9\\13\\17\end{pmatrix}$

`sum` is broadcasted to match the dimensions of `z`:

$\text{broadcasted sum} = \begin{pmatrix}1 && 1\\5 && 5\\9 && 9\\13 && 13\\17 && 17\end{pmatrix}$

And finally,

$\mathbf{z} / \text{broadcasted sum} = \begin{pmatrix}0/1 && 1/1\\2/5 && 3/5\\4/9 && 5/9\\6/13 && 7/13\\8/17 && 9/17\end{pmatrix} = \begin{pmatrix}0 && 1\\0.4 && 0.6\\0.4444 && 0.5556\\0.4615 && 0.5385\\0.4706 && 0.5294\end{pmatrix}$

### Dot Product, Matrix-Vector Product, and Matrix-Matrix Product

Dot product over two vectors $\mathbf{x},\mathbf{y} \in \mathbb{R}^d$ is $\mathbf{x}^T\mathbf{y} = \mathbf{y}^T\mathbf{x}=\langle \mathbf{x},\mathbf{y} \rangle = \sum_{i=0}^d \mathbf{x}_i\mathbf{y}_i$

A Matrix-Vector product $\mathbf{A}\mathbf{x}$ is a matrix comprising as rows the dot product of each row of $\mathbf{A}$ with $\mathbf{x}$.

So, if $\mathbf{A} \in \mathbb{R}^{m\times n}= \begin{pmatrix}\mathbf{a}_1^\top\\\mathbf{a}_2^\top\\\vdots\\\mathbf{a}_m^\top\end{pmatrix}$, where $\mathbf{a}_1, \mathbf{a}_2, \dots, \mathbf{a}_m$ are column vectors, and $\mathbf{x}\in \mathbb{R}^{n} = \begin{pmatrix}\mathbf{x}_1\\\mathbf{x}_2\\\vdots\\\mathbf{x}_n\end{pmatrix}$ is a column vector, then $\mathbf{Ax}\in \mathbb{R}^{m} = \begin{pmatrix}\mathbf{a}_1^\top\mathbf{x}\\\mathbf{a}_2^\top\mathbf{x}\\\vdots\\\mathbf{a}_m^\top\mathbf{x}\end{pmatrix}$

A Matrix-Matrix product is the above idea extended to additional dimensions.

So, if $\mathbf{A} \in \mathbb{R}^{m\times n}= \begin{pmatrix}\mathbf{a}_1^\top\\\mathbf{a}_2^\top\\\vdots\\\mathbf{a}_m^\top\end{pmatrix}$, where $\mathbf{a}_1, \mathbf{a}_2, \dots, \mathbf{a}_m$ are column vectors, and $\mathbf{B} \in \mathbb{R}^{n\times k}= \begin{pmatrix}\mathbf{b}_1 && \mathbf{b}_2 && \dots && \mathbf{b}_k\end{pmatrix}$, where $\mathbf{b}_1, \mathbf{b}_2, \dots, \mathbf{b}_k$ are column vectors, then $\mathbf{AB} \in \mathbb{R}^{m\times k}= \begin{pmatrix}\mathbf{a}_1^\top\mathbf{b}_1 && \mathbf{a}_1^\top\mathbf{b}_2 &&\dots && \mathbf{a}_1^\top\mathbf{b}_k \\\mathbf{a}_2^\top\mathbf{b}_1 && \mathbf{a}_2^\top\mathbf{b}_2 &&\dots && \mathbf{a}_2^\top\mathbf{b}_k \\\vdots && \vdots && \ddots && \vdots\\\mathbf{a}_m^\top\mathbf{b}_1 && \mathbf{a}_m^\top\mathbf{b}_2 &&\dots && \mathbf{a}_m^\top\mathbf{b}_k \\\end{pmatrix}$

In [ ]:
x = torch.tensor([1., 2., 0.])
y = torch.ones(3) * 2
print("x:", x)
print("y:", y)
print("dot product:", torch.dot(x, y))
print("dot product (second way):", torch.sum(x * y))

A = torch.arange(6, dtype = torch.float32).reshape(2,3)
print("\nA:\n", A)
print("x:", x)
print("matrix-vector product:\n", torch.mv(A, x))
print("matrix-vector product: (second way)\n", A@x)

A = torch.arange(6, dtype = torch.float32).reshape(2,3)
B = A.T
print("\nA:\n", A)
print("B:\n", B)
print("matrix-matrix product: \n", A@B)

x: tensor([1., 2., 0.])
y: tensor([2., 2., 2.])
dot product: tensor(6.)
dot product (second way): tensor(6.)

A:
 tensor([[0., 1., 2.],
        [3., 4., 5.]])
x: tensor([1., 2., 0.])
matrix-vector product:
 tensor([ 2., 11.])
matrix-vector product: (second way)
 tensor([ 2., 11.])

A:
 tensor([[0., 1., 2.],
        [3., 4., 5.]])
B:
 tensor([[0., 3.],
        [1., 4.],
        [2., 5.]])
matrix-matrix product: 
 tensor([[ 5., 14.],
        [14., 50.]])


### Norms

$l_p$ norm: $\|\mathbf{x}\|_p=\left( \sum_{i=0}^{n} |x_i|^p\right)^{1/p}$

A familiar form of this is when $p=2$, which is called the $l_2$ norm, which denotes the Euclidean length of a vector:  
$\|\mathbf{x}\|_2=\left( \sum_{i=0}^{n} x_i^2\right)^{1/2}$

Another common one is when $p=1$, which is called the $l_1$ norm, which calculates the Manhattan distance:  
$\|\mathbf{x}\|_1= \sum_{i=0}^{n} |x_i|$

In [ ]:
g = torch.tensor([3., -4.])
print("g:", g)
print("l2 norm:", torch.norm(g))
print("l1 norm:", torch.abs(g).sum())

g: tensor([ 3., -4.])
l2 norm: tensor(5.)
l1 norm: tensor(7.)


## Calculus

### Derivatives and Differentiation

The derivative of a function $f$ at point $x$ is given by $f'(x) = \lim_{h \to 0}\frac{f(x+h)-f(x)}{h}$. It is the instantaneous rate of change of $f$ with respect to $x$.

When $f'(x)$ exists for all $x$ in a given interval $[a, b]$, then $f$ is said to be differentiable on this set.



In [ ]:
def f(x):
    return 3 * x ** 2 - 4 * x

def f_prime(x):
    return 6 * x - 4

print("f'(1):", f_prime(1))

print("\nFor x = 1, as h approaches 0:")
for h in 10.0**torch.arange(-1, -6, -1):
    print(f'h={h:.5f}, numerical limit={(f(1+h)-f(1))/h:.5f}')

f'(1): 2

For x = 1, as h approaches 0:
h=0.10000, numerical limit=2.30000
h=0.01000, numerical limit=2.02999
h=0.00100, numerical limit=2.00295
h=0.00010, numerical limit=2.00033
h=0.00001, numerical limit=2.00272


### Common Derivatives

- $\frac{d}{dx}C = 0$ for any constant C
- $\frac{d}{dx}x^n = nx^{n-1}$ for $n \ne 0$
- $\frac{d}{dx}e^x = e^x$
- $\frac{d}{dx}\log(x) = \frac{1}{x}$

### Properties of Derivatives

- Constant Multiple Rule  
$\frac{d}{dx}[C f(x)] = C\frac{d}{dx}f(x)$ for any constant C  
- Sum Rule  
$\frac{d}{dx}[f(x) + g(x)] = \frac{d}{dx}f(x) + \frac{d}{dx}g(x)$  
- Product Rule  
$\frac{d}{dx}[f(x)g(x)] = f(x)\frac{d}{dx}g(x) + g(x)\frac{d}{dx}f(x)$
- Quotient Rule  
$\frac{d}{dx}\left[\frac{f(x)}{g(x)}\right] = \frac{g(x)\frac{d}{dx}f(x) - f(x)\frac{d}{dx}g(x)}{g(x)^2}$

Quick demo:

If $f(x) = x\log(x)$, then  
$f'(x) = x \left(\frac{d}{dx} \log(x)\right) + \left(\frac{d}{dx}x\right) \log(x) = x (1/x) + 1.\log(x) = 1 + \log(x)$

### Partial Derivatives and Gradients

Extending the above to multivariate functions,

$\frac{\partial f}{\partial x_i} = \partial_{x_i}f = \lim_{h \to 0} \frac{f(x_1, \dots, x_{i-1}, x_i + h, x_{i+1}, \dots, x_n) - f(x_1, \dots, x_i, \dots, x_n)}{h}$

Gradient of function $f$ with respect to $x$ is a vector of $n$ partial derivatives:

$\nabla_x f(x) = [\partial_{x_1}f, \partial_{x_2}f, \dots, \partial_{x_n}f]^\top$


### Chain Rule

To help with calculating the gradients of deeply nested functions (functions of functions of ...), we use the chain rule.

In the univariate case, if $y = f(u)$ and $u = g(x)$, then
$\frac{dy}{dx} = \frac{df}{du}\frac{du}{dx}$

For example,  
$\frac{d\log(x^2)}{dx} = \frac{d\log(x^2)}{dx^2}\frac{dx^2}{dx} = \frac{1}{x^2}\cdot 2x = \frac{2}{x}$

In the multivariate case, if $y = f(u_1, u_2, \dots, u_m)$ and $u_i = g_i(x_1, x_2, \dots, x_n)$, then  
$\frac{dy}{dx_i} = \frac{df}{du_1}\frac{du_1}{dx_i} + \frac{df}{du_2}\frac{du_2}{dx_i} + \dots + \frac{df}{du_m}\frac{du_m}{dx_i}$

For example,  
If $z(x, y) = \log(x + y)$, where $x(t) = t^2$ and $y(t) = t$, then  
$\frac{dz}{dt} = \frac{dz}{dx}\frac{dx}{dt} + \frac{dz}{dy}\frac{dy}{dt} = \frac{1}{x}\cdot 2t + \frac{1}{y}\cdot 1 = \frac{1}{t^2}\cdot 2t + \frac{1}{t} = \frac{2}{t} + \frac{1}{t} = \frac{3}{t}$